packages/kaiju-mapper/src/kaiju_mapper/gmapper.py

```diff
@@ -0,0 +1,438 @@
+                    )
+                    cover.intervals = new_intervals
+
+                    del ad_scores[best_split]
+                    del split_index[best_split]
+
+            elif self.method == "randomized":
+                all_elements_idx = [i for i in range(len(cover.intervals))]
+                element_ad_scores = [
+                    self.ad_test(interval_membership[i])
+                    for i in range(len(cover.intervals))
+                ]
+                element_ad_scores = [
+                    0 if x != x else x for x in element_ad_scores
+                ]  # Handle NaN
+
+                if sum(element_ad_scores) == 0:
+                    logger.info(
+                        f"Convergence after {iteration} iterations - all AD scores are zero."
+                    )
+                    return cover.intervals
+
+                found_valid = False
+                while not found_valid and len(all_elements_idx) > 0:
+                    # Sample one of the intervals weighted by ad score
+                    weights = np.asarray(element_ad_scores)[all_elements_idx]
+                    if weights.sum() == 0:
+                        # If all weights are zero, use uniform weights
+                        weights = np.ones_like(weights) / len(weights)
+                    else:
+                        weights = weights / weights.sum()
+
+                    current_element = int(
+                        np.random.choice(np.asarray(all_elements_idx), p=weights)
+                    )
+                    j = current_element
+
+                    if len(interval_membership[j]) == 0:
+                        removal_idx = all_elements_idx.index(j)
+                        all_elements_idx.pop(removal_idx)
+                        continue
+
+                    if self.ad_test(interval_membership[j]) > self.ad_threshold:
+                        check_interval[j] = True
+                        check_interval.insert(j + 1, True)
+
+                        new_intervals, interval_membership = self._split(
+                            interval_membership, cover.intervals, self.g_overlap, j
+                        )
+                        cover.intervals = new_intervals
+
+                        found_valid = True
+                    else:
+                        removal_idx = all_elements_idx.index(j)
+                        all_elements_idx.pop(removal_idx)
+
+                if not found_valid:
+                    logger.info(
+                        f"Convergence after {iteration} iterations - no valid splits found."
+                    )
+                    return cover.intervals
+
+            if len(cover.intervals) > self.max_intervals:
+                logger.info(
+                    f"Reached maximum number of intervals ({self.max_intervals})."
+                )
+                break
+
+        return cover.intervals
```

Comment:
This 200 line block was basically impossible for me to follow. I would personally break it into one method for each strategy and go from there. Something like b9c610d7a65af1ff8de999d4dafeedb5b26a4dca.